plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -153,213 +200,263 @@ public LocationMode getLocationMode() {
+                }
+            });
+        } catch (Exception e) {
+            throw new IOException("Unable to delete blobs " + blobs, e);
         }
-        return new DeleteResult(blobsDeleted.get(), bytesDeleted.get());
+
+        deleteBlobsInBatches(blobURLs);
     }
 
-    public InputStream getInputStream(String blob, long position, @Nullable Long length) throws URISyntaxException, StorageException {
-        final Tuple<CloudBlobClient, Supplier<OperationContext>> client = client();
-        final OperationContext context = hookMetricCollector(client.v2().get(), getMetricsCollector);
-        final CloudBlockBlob blockBlobReference = client.v1().getContainerReference(container).getBlockBlobReference(blob);
-        logger.trace(() -> new ParameterizedMessage("reading container [{}], blob [{}]", container, blob));
-        final long limit;
-        if (length == null) {
-            // Loading the blob attributes so we can get its length
-            SocketAccess.doPrivilegedVoidException(() -> blockBlobReference.downloadAttributes(null, null, context));
-            limit = blockBlobReference.getProperties().getLength() - position;
+    private void deleteBlobsInBatches(List<String> blobUrls) throws IOException {
+        if (blobUrls.isEmpty()) {
+            return;
         }
-        else {
-            limit = length;
-        }
-        final BlobInputStream blobInputStream = new BlobInputStream(limit, blockBlobReference, position, context);
-        if (length != null) {
-            // pre-filling the buffer in case of ranged reads so this method throws a 404 storage exception right away in case the blob
-            // does not exist
-            blobInputStream.fill();
+
+        try {
+            SocketAccess.doPrivilegedVoidException(() -> {
+                final BlobBatchAsyncClient blobBatchClient =
+                    new BlobBatchClientBuilder(asyncClient())
+                        .buildAsyncClient();
+
+                int numBatches = (int) Math.ceil((double) blobUrls.size() / (double) MAX_ELEMENTS_PER_BATCH);
+                List<BlobBatch> batches = new ArrayList<>(numBatches);
+                for (int batchNumber = 0; batchNumber < numBatches; batchNumber++) {
+                    final BlobBatch blobBatch = blobBatchClient.getBlobBatch();
+                    int rangeStart = batchNumber * MAX_ELEMENTS_PER_BATCH;
+                    for (int i = rangeStart; i < Math.min(rangeStart + MAX_ELEMENTS_PER_BATCH, blobUrls.size()); i++) {
+                        blobBatch.deleteBlob(blobUrls.get(i));
+                    }
+                    batches.add(blobBatch);
+                }
+
+                List<Mono<Response<Void>>> batchResponses = new ArrayList<>(batches.size());
+                for (BlobBatch batch : batches) {
+                    batchResponses.add(blobBatchClient.submitBatchWithResponse(batch, false));
```

Comment:
You're right, I've just tested this using a SAS token and I got a:
```
com.azure.storage.blob.models.BlobStorageException: Status code 403, "<?xml version="1.0" encoding="utf-8"?><Error><Code>AuthorizationResourceTypeMismatch</Code><Message>This request is not authorized to perform this operation using this resource type.
```

I'll implement a different path that triggers all the delete requests when a SAS token is used :(.

Reply:
> I'll implement a different path that triggers all the delete requests when a SAS token is used :(.

Let's just use a single path IMO. No need to optimize for non-token credentials, those are pretty rare in production.
I think we can still get into a much better state than what we had with the v8 SDK because we can just fire off all the deletes at once with the async API?

Reply:
👍, yes that was my plan.